lib/maths/CBoostedTreeLeafNodeStatistics.cc

```diff
@@ -267,6 +335,20 @@ void CBoostedTreeLeafNodeStatistics::addRowDerivatives(const CEncodedDataFrameRo
 
     auto derivatives = readLossDerivatives(row.unencodedRow(), m_ExtraColumns,
                                            m_NumberLossParameters);
+    if (derivatives.size() == 2) {
```

Comment:
This is the hottest path in the training code. Whilst this condition is predictable we know outside of this function if it true or not. It might be worth having another version of this function, you could have say `updateDerivatives` and `updateDerivativesAndBounds`, and pulling the check out of the loop over rows.

Reply:
I tried this change in there was so difference in the performance :man_shrugging: I guess the compiler is already good enough to restructure/optimize the loop. Therefore, I will not create two versions of the function to avoid code duplication.

Reply:
Done